Swappable Site Model

silkentrance opened the pull request:
fixes #4

- make site model swappable

matthiask commented on feincms3_sites/__init__.py:
```diff
@@ -1,4 +1,24 @@
+from django.apps import apps as django_apps
+from django.conf import settings
+from django.core.exceptions import ImproperlyConfigured
```
Can you move those imports into the `get_site_model()` function? My (possibly archaic) packaging workflow still requires that `setup.py` is allowed to import `feincms3_sites/__init__.py`

silkentrance replied:
Uh, I took that over from auth and tried out different things including also having it in `feincms3_sites.models`. Will move it back to `feincms3_sites.models`.

`get_site_model` is a convenience function for users who want to associate custom (app) content with specific sites.

Inside `feincms3_sites.model` we cannot use that function as it will fail since the other app and its models may not be initialized, hence the signal handler in `feincms3_sites.models` for the `AbstractPage`.

silkentrance replied:
models are partially initialized, will have to disect the models module into a models package

silkentrance replied:
i had to move `get_site_model` to `models/utils.py` to overcome circular depencendies and also premature import of otherwise uninitialized/unprepared models.

matthiask replied:
I'd rather have the parts be in one file than introduce a models package if it is possible. This was just a suggestion, I'm sorry for the trouble it caused 😅

silkentrance replied:
hm, get_site_model is used by the middleware, which cannot include the models as they are only partially initialized at the time that the middleware is realized.
so i need to move that function out of the models in order to make the models a single package again, say `feincms3_sites.utils`.

silkentrance replied:
refactored.

silkentrance replied:
but not tested yet :-)

silkentrance replied:
well. it works...

matthiask commented on feincms3_sites/middleware.py:
```diff
@@ -23,10 +25,9 @@ def current_site():
-
     def middleware(request):
-        request.site = Site.objects.for_host(request.get_host())
+        site_model = get_site_model()
```
Django seems to fetch the user model when importing modules, see e.g. https://github.com/django/django/blob/c5776bfca9e3f35e0ab5aacbdc1a4dbfe92fdfd1/django/contrib/auth/backends.py#L5

Would that be possible here as well?

silkentrance replied:
This does not work as the queryset derived manager from `AbstractSite` is bound to the model class that it was registered with, reverting this back gives me the site model was swapped for error and that the manager cannot be used with that.

matthiask replied:
Makes sense.

matthiask commented on feincms3_sites/models.py:
```diff
@@ -28,7 +28,7 @@ def for_host(self, host):
 
 
-class Site(models.Model):
+class AbstractSite(models.Model):
```
django.contrib.auth uses a `base_user` module for its abstract models to avoid importing the `models` file containing the swapped out version if it is not required. Maybe that's not necessary, but it may be nice. What do you think?

silkentrance replied:
This would be overly complex for the site model, as the base is already the AbstractSite and everything in that module, i.e. `site` already declares/defines everything that is needed for a basic site.

The user model is rather complex as it deals with security and the site model does not need such complexity as the user model already provides for that. What the site model needs, though, are additional fields that can then be rendered in the templates or used in helper functions to create for example meta data and whatnot.

So I would keep it rather simple and not introduce another level of abstraction.

matthiask replied:
Oh I think I didn't express myself clearly. I didn't want to add any authentication or security, only suggest that it may be easier when separating the abstract models into a `feincms3_sites/base.py` module and keep the concrete model in `feincms3_sites/models.py` or something. 

It probably doesn't matter.

silkentrance replied:
Of course not, I just wanted to point out that the requirements for the user model are somewhat more elaborate and that splitting the models of feincms3_sites into separate basemodels and models modules will provide no extra benefit.

matthiask replied:
As long as Django doesn't complain about importing models which aren't migrated or part of `INSTALLED_APPS` that's fine. I'm not sure but I remember that there was trouble in this area once. 

Anyway, if it works it works and that's definitely enough for me .)

matthiask commented on feincms3_sites/models.py:
```diff
@@ -150,3 +154,13 @@ def save(self, *args, **kwargs):
+    @staticmethod
+    def add_site_field(sender, **kwargs):
+        model = settings.FEINCMS3_SITES_SITE_MODEL
+        if (sender._meta.app_label + '.' + sender._meta.object_name) == model:
```
`sender._meta.label_lower == model.lower()` or something like that? I think you shouldn't depend on the exact capitalization of model names.

silkentrance replied:
Yes!

silkentrance replied:
done

matthiask reviewed (commented):
I didn't take a look at the test failures (for now). You should definitely run `tox -e style` once to format the code according to black's standards etc.

Apart from the comments I agree with the changes, looks good!

silkentrance commented:
> I didn't take a look at the test failures (for now). You should definitely run `tox -e style` once to format the code according to black's standards etc.

Well, I didn't either, as they do not concern me right now. No tests have been implemented, yet.

As soon as we come to an overall conclusion, I will implement the required tests.

silkentrance commented:
the failing tests resolve to feincms3 ckeditor.
can you fix these so that the non-existing tests for this will yield a positive result?

matthiask commented:
I have released feincms3==0.94.1 which works with Python < 3.8 again and'll trigger a new build.

silkentrance commented:
`:=` is `Python 3.8+`

matthiask commented:
Yes, I've fixed this a few days ago in feincms3 but I haven't uploaded a patch release containing this fix.

silkentrance commented:
> Yes, I've fixed this a few days ago in feincms3 but I haven't uploaded a patch release containing this fix.

I was just about to adjust setup.cfg to make sure that python>=3.8 is required.

silkentrance commented:
Okay, fixed the issue with the site foreign key being added to string instead of the sender. Eliminated duplicate code.
Added FEINCMS3_SITES_SITE_MODEL to testapp settings as django's orm will check that key on start.

silkentrance commented:
removed the extraneous init module from root (was added by pycharm, argh)
removed the duplicate code.

matthiask commented:
Nice, thanks!
